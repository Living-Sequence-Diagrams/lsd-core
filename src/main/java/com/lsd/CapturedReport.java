package com.lsd;

import lombok.Value;

@Value
public class CapturedReport {
    String title;
    String filename;
    String status;
}
