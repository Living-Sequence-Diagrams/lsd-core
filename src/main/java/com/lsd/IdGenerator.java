package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
