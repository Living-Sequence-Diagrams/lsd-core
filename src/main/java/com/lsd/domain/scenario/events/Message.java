package com.lsd.domain.scenario.events;

import lombok.Builder;
import lombok.Value;

import static java.lang.String.format;

@Value
@Builder
public class Message implements Interaction {
    String id;
    String label;
    String from;
    String to;
    Object data;

    public String toMarkup() {
        return format("%s -> %s:[[#%s %s]]",
                getFrom(), getTo(), getId(), getLabel());
    }
}
