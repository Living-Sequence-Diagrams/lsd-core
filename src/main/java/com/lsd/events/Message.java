package com.lsd.events;

import com.lsd.report.model.DataHolder;
import lombok.Builder.Default;
import lombok.Data;
import lombok.experimental.SuperBuilder;
import org.apache.commons.text.StringSubstitutor;

import java.util.Map;

import static com.lsd.events.ArrowType.SOLID;

@Data
@SuperBuilder
public class Message implements DataHolder {
    private final String id;
    private final String from;
    private final String to;
    private final String label;

    @Default
    private final Object data = "";

    @Default
    private final String colour = "";

    @Default
    private final ArrowType arrowType = SOLID;

    public String toMarkup() {
        return StringSubstitutor.replace("${from} ${arrow} ${to}: <text fill=\"${colour}\">[[#${id} {${tooltip}} ${label}]]</text>", Map.of(
                "from", getFrom(),
                "to", getTo(),
                "id", getId(),
                "tooltip", getLabel().replaceAll("<\\$.*?>", ""),
                "label", abbreviatedLabel(),
                "colour", getColour(),
                "arrow", getArrowType().toMarkup(getColour())
        ));
    }

}
