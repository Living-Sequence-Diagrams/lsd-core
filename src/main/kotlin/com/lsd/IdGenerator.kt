package com.lsd

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    
    private val idSupplier: Supplier<String> = if (isDeterministic)
        Supplier { counter.incrementAndGet().toString() }
    else
        Supplier { randomUUID().toString().replace("-".toRegex(), "") }


    operator fun next(): String = idSupplier.get()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}