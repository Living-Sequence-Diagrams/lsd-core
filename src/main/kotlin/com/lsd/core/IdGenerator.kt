package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}