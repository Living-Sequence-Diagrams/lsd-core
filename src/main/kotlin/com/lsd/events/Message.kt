package com.lsd.events

import com.lsd.builders.MessageBuilder
import com.lsd.events.ArrowType.SOLID

data class Message(
    override val id: String,
    override val label: String,
    override val data: Any? = null,
    override val from: String,
    override val to: String,
    override val colour: String = "",
    override val arrowType: ArrowType = SOLID,
) : SequenceMessage {
    companion object {
        @JvmStatic
        fun builder(): MessageBuilder = MessageBuilder()
    }
}