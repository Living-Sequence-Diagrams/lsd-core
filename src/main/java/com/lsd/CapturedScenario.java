package com.lsd;

import com.lsd.events.SequenceEvent;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

import static java.util.Objects.nonNull;

@NoArgsConstructor
@Data
public class CapturedScenario {
    private final List<SequenceEvent> sequenceEvents = new ArrayList<>();

    private String title;
    private String description;

    public void add(SequenceEvent sequenceEvent) {
        if (nonNull(sequenceEvent)) {
            sequenceEvents.add(sequenceEvent);
        }
    }
}
