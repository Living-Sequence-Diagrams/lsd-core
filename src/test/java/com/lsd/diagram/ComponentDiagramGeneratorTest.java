package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.Message;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.lsd.ParticipantType.ACTOR;
import static com.lsd.ParticipantType.PARTICIPANT;

class ComponentDiagramGeneratorTest {

    private final IdGenerator idGenerator = new IdGenerator(true);

    @Test
    void removesDuplicateInteractions() {
        var sampleMessage = sampleMessage();
        var diagramGenerator = new ComponentDiagramGenerator(List.of(), List.of(sampleMessage, sampleMessage, sampleMessage), idGenerator);
        var diagram = diagramGenerator.diagram().orElseThrow();

        Approvals.verify(diagram.getUml());
    }

    @Test
    void removesDuplicateParticipants() {
        var sampleParticipant = ACTOR.called("Nick");
        var diagramGenerator = new ComponentDiagramGenerator(List.of(sampleParticipant, sampleParticipant), List.of(sampleMessage()), idGenerator);
        var diagram = diagramGenerator.diagram().orElseThrow();

        Approvals.verify(diagram.getUml());
    }

    @Test
    void replacesParticipantKeywordWithComponent() {
        var sampleParticipant = PARTICIPANT.called("SomeService");
        var diagramGenerator = new ComponentDiagramGenerator(List.of(sampleParticipant), List.of(sampleMessage()), idGenerator);
        var diagram = diagramGenerator.diagram().orElseThrow();

        Approvals.verify(diagram.getUml());
    }

    private Message sampleMessage() {
        return Message.builder()
                .from("A")
                .to("B")
                .build();
    }
}