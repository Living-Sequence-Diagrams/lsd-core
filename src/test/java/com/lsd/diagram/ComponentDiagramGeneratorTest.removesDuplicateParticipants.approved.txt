@startuml

!theme plain
skinparam backgroundColor transparent

actor Nick

[A] -> B

hide @unlinked
@enduml